From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Sat, 6 Apr 2024 22:57:41 -0400
Subject: [PATCH] Optimize Minecart collisions

Co-authored-by: MrHua269 <wangxyper@163.com>

Skip tick collisions to to prevent lag causing by massive stacked Minecart
Useful for anarchy server.

diff --git a/src/main/java/net/minecraft/world/entity/EntitySelector.java b/src/main/java/net/minecraft/world/entity/EntitySelector.java
index cfa9607241c3e69777ffc317206996c2f783437a..fa8aaa5c30e53aa98aab15a924e27df92af41f2d 100644
--- a/src/main/java/net/minecraft/world/entity/EntitySelector.java
+++ b/src/main/java/net/minecraft/world/entity/EntitySelector.java
@@ -63,6 +63,13 @@ public final class EntitySelector {
 
     public static Predicate<Entity> pushable(Entity entity, boolean ignoreClimbing) {
         // Paper end
+
+        // Leaf start - Optimize Minecart collisions
+        if (entity instanceof net.minecraft.world.entity.vehicle.AbstractMinecart) {
+            return x -> true;
+        }
+        // Leaf end - Optimize Minecart collisions
+
         Team scoreboardteambase = entity.getTeam();
         Team.CollisionRule scoreboardteambase_enumteampush = scoreboardteambase == null ? Team.CollisionRule.ALWAYS : scoreboardteambase.getCollisionRule();
 
diff --git a/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java b/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java
index c9b7b12470afac45b0132858407aacb8f91aac68..86df284cd88a20779a8aec2daec02a9dd1dc69d8 100644
--- a/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java
+++ b/src/main/java/net/minecraft/world/entity/vehicle/AbstractMinecart.java
@@ -439,13 +439,15 @@ public abstract class AbstractMinecart extends Entity {
                 this.level().getCraftServer().getPluginManager().callEvent(new org.bukkit.event.vehicle.VehicleMoveEvent(vehicle, from, to));
             }
             // CraftBukkit end
-            if (this.getMinecartType() == AbstractMinecart.Type.RIDEABLE && this.getDeltaMovement().horizontalDistanceSqr() > 0.01D) {
+            // Leaf start - Optimize Minecart collision handling
+            // The logic below is used to get list of entities around Minecart
+            // and handle behaviors for their collisions with each other
+            if (!org.dreeam.leaf.config.modules.opt.OptimizeMinecart.enabled || this.tickCount % org.dreeam.leaf.config.modules.opt.OptimizeMinecart.skipTickCount == 0) {
+            if (this.getMinecartType() == AbstractMinecart.Type.RIDEABLE && (org.dreeam.leaf.config.modules.opt.OptimizeMinecart.enabled || this.getDeltaMovement().horizontalDistanceSqr() > 0.01D)) {
                 List<Entity> list = this.level().getEntities((Entity) this, this.getBoundingBox().inflate(0.20000000298023224D, 0.0D, 0.20000000298023224D), EntitySelector.pushableBy(this));
 
                 if (!list.isEmpty()) {
-                    for (int l = 0; l < list.size(); ++l) {
-                        Entity entity = (Entity) list.get(l);
-
+                    for (Entity entity : list) {
                         if (!(entity instanceof Player) && !(entity instanceof IronGolem) && !(entity instanceof AbstractMinecart) && !this.isVehicle() && !entity.isPassenger()) {
                             // CraftBukkit start
                             VehicleEntityCollisionEvent collisionEvent = new VehicleEntityCollisionEvent(vehicle, entity.getBukkitEntity());
@@ -472,11 +474,7 @@ public abstract class AbstractMinecart extends Entity {
                     }
                 }
             } else {
-                Iterator iterator = this.level().getEntities(this, this.getBoundingBox().inflate(0.20000000298023224D, 0.0D, 0.20000000298023224D)).iterator();
-
-                while (iterator.hasNext()) {
-                    Entity entity1 = (Entity) iterator.next();
-
+                for (Entity entity1 : this.level().getEntities(this, this.getBoundingBox().inflate(0.20000000298023224D, 0.0D, 0.20000000298023224D))) {
                     if (!this.hasPassenger(entity1) && entity1.isPushable() && entity1 instanceof AbstractMinecart) {
                         // CraftBukkit start
                         VehicleEntityCollisionEvent collisionEvent = new VehicleEntityCollisionEvent(vehicle, entity1.getBukkitEntity());
@@ -490,6 +488,8 @@ public abstract class AbstractMinecart extends Entity {
                     }
                 }
             }
+            }
+            // Leaf end
 
             this.updateInWaterStateAndDoFluidPushing();
             if (this.isInLava()) {
diff --git a/src/main/java/net/minecraft/world/level/Level.java b/src/main/java/net/minecraft/world/level/Level.java
index 78f2bea7ae34be7a2152d7c31e90a524eac1ddfe..00b661973b73d7a75277511cdf8e7efe1311ef42 100644
--- a/src/main/java/net/minecraft/world/level/Level.java
+++ b/src/main/java/net/minecraft/world/level/Level.java
@@ -1238,12 +1238,14 @@ public abstract class Level implements LevelAccessor, AutoCloseable {
         return this.getChunk(chunkX, chunkZ, ChunkStatus.FULL, false);
     }
 
+    // Leaf start - Optimize predicate call
     @Override
-    public List<Entity> getEntities(@Nullable Entity except, AABB box, Predicate<? super Entity> predicate) {
+    public List<Entity> getEntities(@Nullable Entity except, AABB box, @Nullable Predicate<? super Entity> predicate) {
         List<Entity> list = Lists.newArrayList();
-        ((ServerLevel)this).getEntityLookup().getEntities(except, box, list, predicate); // Paper - optimise this call
-        return list;
+        ((ServerLevel) this).getEntityLookup().getEntities(except, box, list, null); // Paper - optimise this call
+        return predicate != null ? list.stream().filter(predicate).toList() : list;
     }
+    // Leaf end
 
     @Override
     public <T extends Entity> List<T> getEntities(EntityTypeTest<Entity, T> filter, AABB box, Predicate<? super T> predicate) {
diff --git a/src/main/java/org/dreeam/leaf/config/modules/opt/OptimizeMinecart.java b/src/main/java/org/dreeam/leaf/config/modules/opt/OptimizeMinecart.java
new file mode 100644
index 0000000000000000000000000000000000000000..a891454eb9b83d50ef3e8914efbb013f88a6b3fd
--- /dev/null
+++ b/src/main/java/org/dreeam/leaf/config/modules/opt/OptimizeMinecart.java
@@ -0,0 +1,28 @@
+package org.dreeam.leaf.config.modules.opt;
+
+import org.dreeam.leaf.config.ConfigInfo;
+import org.dreeam.leaf.config.EnumConfigCategory;
+import org.dreeam.leaf.config.IConfigModule;
+
+public class OptimizeMinecart implements IConfigModule {
+
+    @Override
+    public EnumConfigCategory getCategory() {
+        return EnumConfigCategory.PERFORMANCE;
+    }
+
+    @Override
+    public String getBaseName() {
+        return "optimize_minecart";
+    }
+
+    @ConfigInfo(baseName = "enabled", comments = """
+            Enable this feature to handle large amount of stacked Minecart better.
+            By skipping tick collisions to reduce expense getting entities list 
+            and bukkit event calls, useful for the anarchy server.
+            """)
+    public static boolean enabled = false;
+
+    @ConfigInfo(baseName = "skip-tick-count")
+    public static int skipTickCount = 30;
+}
